x-pack/solutions/observability/plugins/apm/common/service_map/get_service_map_nodes.ts

```diff
@@ -0,0 +1,295 @@
+  );
+
+  // 1. Map external nodes to internal services
+  // 2. Collapse external nodes into one node based on span.destination.service.resource
+  // 3. Pick the first available span.type/span.subtype in an alphabetically sorted list
+  return Array.from(nodes.entries()).reduce((map, [id, node]) => {
+    if (map.has(id)) {
+      return map;
+    }
+
+    const isMessagingSpan = messagingSpanIds.has(node.id);
+    const destinationService = isMessagingSpan ? undefined : exitSpanDestinationsMap.get(node.id);
+    const isServiceNode = !!destinationService || !isExitSpan(node);
+
+    if (isServiceNode) {
+      const serviceId = destinationService ? destinationService.id : node.id;
+      const serviceNode = services.get(serviceId);
+
+      if (serviceNode) {
+        const serviceAnomalyStats = anomaliesByServiceName.get(serviceNode.id);
+
+        map.set(node.id, {
+          ...serviceNode,
+          ...(serviceAnomalyStats ? { serviceAnomalyStats } : null),
+        });
+      }
+    } else {
+      const exitSpanNodes = exitSpans.get(id) ?? [];
+      if (exitSpanNodes.length > 0) {
+        const exitSpanSample = exitSpanNodes[0];
+        map.set(id, {
+          ...exitSpanSample,
+          label: exitSpanSample[SPAN_DESTINATION_SERVICE_RESOURCE],
+          [SPAN_TYPE]: exitSpanNodes.map((n) => n[SPAN_TYPE]).sort()[0],
+          [SPAN_SUBTYPE]: exitSpanNodes.map((n) => n[SPAN_SUBTYPE]).sort()[0],
+        });
+      }
+    }
+
+    return map;
+  }, new Map<string, ConnectionNode>());
+}
+
+function mapEdges({
+  allConnections,
+  nodes,
+}: {
+  allConnections: Connection[];
+  nodes: Map<string, ConnectionNode>;
+}) {
+  const connections = allConnections.reduce((acc, connection) => {
+    const sourceData = nodes.get(connection.source.id);
+    const targetData = nodes.get(connection.destination.id);
+
+    if (!sourceData || !targetData || sourceData.id === targetData.id) {
+      return acc;
+    }
+
+    const label = `${sourceData[SERVICE_NAME]} to ${
+      targetData[SERVICE_NAME] || targetData[SPAN_DESTINATION_SERVICE_RESOURCE]
+    }`;
+    const id = getConnectionId({ source: sourceData, destination: targetData });
+
+    acc.set(id, {
+      source: sourceData.id,
+      target: targetData.id,
+      label,
+      id,
+      sourceData,
+      targetData,
+    });
+
+    return acc;
+  }, new Map<string, ConnectionEdge & { sourceData: ConnectionNode; targetData: ConnectionNode }>());
+
+  return Array.from(connections.values());
+}
+
+function markBidirectionalConnections({ connections }: { connections: ConnectionEdge[] }) {
+  return connections.reduce<ConnectionEdge[]>((prev, connection) => {
+    const reversedConnection = prev.find(
+      (c) => c.target === connection.source && c.source === connection.target
+    );
+
+    if (reversedConnection) {
+      reversedConnection.bidirectional = true;
+      return prev.concat({
+        ...connection,
```

Comment:
Does this have to be a new object with different property set? Just trying to understand some of the perf impact of some of these operations